docs/object-capabilities/derived-authority.md

```diff
@@ -0,0 +1,138 @@
+This value is the capability that represents the ambient authority given to us by the system.
+
+Here is a program that connects to example.com via TCP on port 80 and quits:
+
+```pony
+use "net"
+
+class MyTCPConnectionNotify is TCPConnectionNotify
+  let _out: OutStream
+
+  new iso create(out: OutStream) =>
+    _out = out
+
+  fun ref connected(conn: TCPConnection ref) =>
+    _out.print("connected")
+    conn.close()
+
+  fun ref connect_failed(conn: TCPConnection ref) =>
+    _out.print("connect_failed")
+
+actor Connect
+  new create(out: OutStream, auth: AmbientAuth) =>
+    TCPConnection(auth, MyTCPConnectionNotify(out), "example.com", "80")
+
+actor Main
+  new create(env: Env) =>
+    try Connect(env.out, env.root as AmbientAuth) end
+```
+
+The `Main` actor authorizes the `Connect` actor by passing the ambient authority token on to it. This token is unforgeable since the `AmbientAuth` constructor is private and the only existing
+instance is provided by the runtime itself.
+
+The `Connect` actor uses this authority when it creates a TCP connection:
+
+```pony
+TCPConnection(auth, MyTCPConnectionNotify(out), "example.com", "80")
+```
+
+The `TCPConnection` requires an authority as first parameter, and since the compiler checks that
+the correct type was passed, this guarantees that a `TCPConnection` can only be created by an
+actor holding the required authorization.
+
+The implementation of the `TCPConnection` constructor does not even use the authorization parameter
+at run time, all it does is require it to be of the right type. The type checking done by the compiler
+is sufficient for this guarantee.
+
+## Restrict, then delegate your authority
+
+In order to handle our own code and that of others more safely, and also to understand our code better,
+we want to split up the authority, and only grant the particular authority a piece of code actually
+requires.
+
+Imagine we don't trust the `Connect` actor, so we don't want to provide it with more authority
+than needed. For example, there is no point in granting it filesystem access, since it is supposed
+to do network things (specifically, TCP), not access the filesystem. To understand how we can achieve this,
+let's look at the authority types involved.
+
+The first parameter of the `TCPConnection` constructor has the type
+
+```pony
+type TCPConnectionAuth is (AmbientAuth | NetAuth | TCPAuth | TCPConnectAuth)
+```
+
+so instead of passing the entire `AmbientAuth`, we can "downgrade" that to a `TCPConnectAuth`,
+pass it to the `Connect` actor and have that pass it to the `TCPConnection` constructor:
+
+```pony
+actor Connect
+  new create(out: OutStream, auth: TCPConnectionAuth) =>
+    TCPConnection(auth, MyTCPConnectionNotify(out), "example.com", "80")
+
+actor Main
+  new create(env: Env) =>
+    try Connect(env.out, TCPConnectAuth(env.root as AmbientAuth)) end
+```
+
+Now we are sure it cannot access the filesystem or listen on a TCP or UDP port.
```

Comment:
This sentence confused me on re-reading as I often forget the hierarchy of authorities in net. I do not think it should block merging this, but perhaps before the code block when talking about "downgrading" a brief little rewrite might help?

> ...`AmbientAuth`, we can "downgrade" that to a `TCPConnectAuth`,...

to

> ...`AmbientAuth` (the root of all authority), we can "downgrade" that to a `TCPConnectAuth` (the most restrictive authority in net),...

Reply:
You explain this further below so not much is needed, but contextualizing the "downgrade" might make it easier to understand on first read.

Reply:
Sure thing!